src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.GroupedAnalyzerActions.cs

```diff
@@ -16,325 +17,65 @@ internal partial class AnalyzerDriver<TLanguageKindEnum> : AnalyzerDriver where
         /// <summary>
         /// <see cref="AnalyzerActions"/> grouped by <see cref="DiagnosticAnalyzer"/>, and possibly other entities, such as <see cref="OperationKind"/>, <see cref="SymbolKind"/>, etc.
         /// </summary>
-        private sealed class GroupedAnalyzerActions
+        private sealed class GroupedAnalyzerActions : IGroupedAnalyzerActions
         {
-            private static readonly GroupedAnalyzerActions s_Default = new GroupedAnalyzerActions(in AnalyzerActions.Empty);
-
-            private readonly AnalyzerActions _analyzerActions;
-
-            private ImmutableDictionary<DiagnosticAnalyzer, ImmutableDictionary<TLanguageKindEnum, ImmutableArray<SyntaxNodeAnalyzerAction<TLanguageKindEnum>>>> _lazyNodeActionsByKind;
-            private ImmutableDictionary<DiagnosticAnalyzer, ImmutableDictionary<OperationKind, ImmutableArray<OperationAnalyzerAction>>> _lazyOperationActionsByKind;
-            private ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<CodeBlockStartAnalyzerAction<TLanguageKindEnum>>> _lazyCodeBlockStartActionsByAnalyzer;
-            // Code block actions and code block end actions are kept separate so that it is easy to
-            // execute the code block actions before the code block end actions.
-            private ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<CodeBlockAnalyzerAction>> _lazyCodeBlockEndActionsByAnalyzer;
-            private ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<CodeBlockAnalyzerAction>> _lazyCodeBlockActionsByAnalyzer;
-            private ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<OperationBlockStartAnalyzerAction>> _lazyOperationBlockStartActionsByAnalyzer;
-            private ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<OperationBlockAnalyzerAction>> _lazyOperationBlockActionsByAnalyzer;
-            private ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<OperationBlockAnalyzerAction>> _lazyOperationBlockEndActionsByAnalyzer;
```

Comment:
All the dictionaries have been removed and the underlying values are not stored in a separate `GroupedAnalyzerActionsForAnalyzer` type. This type now holds an array of tuples `ImmutableArray<(DiagnosticAnalyzer analyzer, GroupedAnalyzerActionsForAnalyzer groupedActions)>`.

Reply:
i'm curious how much win you would have gotten just using a normal dictionary.  i.e. is the impact from using a dict-vs-list?  or is it from immutable-dict in particular (which is *super* inefficient at space, to support easy immtuable updating).

Reply:
Hmm, I am not sure. I just saw the key accesses into the dictionary had lot of CPU stacks in perfview profiles. I figured we don't need any dictionaries and simplified the approach.